### DISCUSSION TITLE
How is pos calculated for TextWriter? The place seems not correct.

### DISCUSSION THREAD

**Discussion by nissansz** [pymupdf/PyMuPDF#3031] [General]
How is pos calculated for TextWriter? The place seems not correct.

            rect = fitz.Rect(left, top, right, bottom)
            tw.append((left, **bottom**), text,  fontsize=fontsize_fit, font=font, small_caps=True)
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/509335b0-62a7-493a-be13-945eaa4c033d)

tw.append((left, **top**), text,  fontsize=fontsize_fit, font=font, small_caps=True)
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/4131ac9c-9fde-4b0f-9101-cdfd6342d801)

tw.append((left, **top+fontsize**), text,  fontsize=fontsize_fit, font=font, small_caps=True)
![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/6afcdb0f-c7da-4842-b222-ae29d3e8ca91)

**Comment by nissansz** [comment]
below setting seems a little better, but don't know whether it is the best option

  fontsize = int((bottom - top)*0.8)
tw.append((left, top+fontsize), text, fontsize=fontsize_fit, font=font, small_caps=True)

![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/a134c9a2-bf0e-4395-9d3d-0319b7ab5170)

**Comment by JorjMcKie** [comment]
You should try to use the **_origin_** of the span that is delivered on text extraction. This value depends on the font. Look at [this](https://pymupdf.readthedocs.io/en/latest/textpage.html#span-dictionary) picture in the documentation.

**Comment by nissansz** [comment]
How to set origin?

a = span["ascender"]
d = span["descender"]
r = fitz.Rect(span["bbox"])
o = fitz.Point(span["origin"])  # its y-value is the baseline
r.y1 = o.y - span["size"] * d / (a - d)
r.y0 = r.y1 - span["size"]
# r now is a rectangle of height 'fontsize'

**Reply by JorjMcKie** [reply]
This is the value you should choose as is for text insertion. There is no other information available.

OCR engines just return the bbox. They do not look at the font. Therefore text extraction has no other way than setting origin to `bbox.bl`.
If text contains characters going below the baseline ("g", "y", "|", ...), this is obviously **_not_** the correct choice. Here, the origin y-value must be chosen a bit upwards.
How much? Unknown - you must guess an ok value.
Your OCR seems to at least tell you the font **_name_**. So you could try to load a similar font, e.g. `font=fitz.Font("tiro")` for Times-Roman. There, you have font.ascender and font.descender. The full character height is `(font.ascender - font.descender) * bbox.height`.
font.descender (negative value) is the part going below the baseline. So `origin = (bbox.x0, bbox.y1 + font.descender * bbox.height)` **_may_** give you a better origin 🤷‍♂️.

**Reply by nissansz** [reply]
I still don't know how to set this origin to TextWriter. 


   tw.append((int(left), int(top+fontsize)), text,  fontsize = fontsize, font=font, small_caps=True)

Change to below?

origin = (bbox.x0, bbox.y1 + font.descender * bbox.height) 
   tw.append(origin , text,  fontsize = fontsize, font=font, small_caps=True)

**Reply by JorjMcKie** [reply]
> origin = (bbox.x0, bbox.y1 + font.descender * bbox.height)

yes, that's my best idea.
**_If_** you have the right font.

**Reply by nissansz** [reply]
result seems wrong.

        origin = (int(left), int(top + font.descender * (bottom-top)))
            tw.append(origin, text,  fontsize = fontsize, font=font, small_caps=True)

![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/1443a945-f203-41f4-a552-70b73f4f1176)

**Reply by JorjMcKie** [reply]
yes, your original calculation is the better approximation - stick with it then.

**Reply by nissansz** [reply]
This seems better 
 fontsize = int((bottom - top)*0.85)
 origin = (int(left), int(bottom+font.descender*(bottom-top)))

![image](https://github.com/pymupdf/PyMuPDF/assets/41010669/6a2eb7bd-0d09-483e-8358-aa2f507f20e3)
